src/util.rs

```diff
@@ -36,42 +36,26 @@ pub fn escape_pango_text(text: String) -> String {
         .collect()
 }
 
-pub fn format_number(raw_value: f64, total_digits: usize, min_unit: &str, suffix: &str) -> String {
-    let (min_unit_value, min_unit_level) = match min_unit {
-        "T" => (raw_value / 1_000_000_000_000.0, 4),
-        "G" => (raw_value / 1_000_000_000.0, 3),
-        "M" => (raw_value / 1_000_000.0, 2),
-        "K" => (raw_value / 1_000.0, 1),
-        "1" => (raw_value, 0),
-        "m" => (raw_value * 1_000.0, -1),
-        "u" => (raw_value * 1_000_000.0, -2),
-        "n" => (raw_value * 1_000_000_000.0, -3),
-        _ => (raw_value * 1_000_000_000_000.0, -4),
+/// Format `raw_value` to engineering notation
+pub fn format_number(raw_value: f64, total_digits: usize, min_suffix: &str, unit: &str) -> String {
+    let min_exp_level = match min_suffix {
+        "T" => 4,
+        "G" => 3,
+        "M" => 2,
+        "K" => 1,
+        "1" => 0,
+        "m" => -1,
+        "u" => -2,
+        "n" => -3,
+        _ => -4,
     };
 
-    //println!("Min Unit:  ({}, {})", min_unit_value, min_unit_level);
-
-    let (magnitude_value, magnitude_level) = match raw_value {
-        x if x >= 100_000_000_000.0 => (raw_value / 1_000_000_000_000.0, 4),
-        x if x >= 100_000_000.0 => (raw_value / 1_000_000_000.0, 3),
-        x if x >= 100_000.0 => (raw_value / 1_000_000.0, 2),
-        x if x >= 100.0 => (raw_value / 1_000.0, 1),
-        x if x >= 0.1 => (raw_value, 0),
-        x if x >= 0.000_1 => (raw_value * 1_000.0, -1),
-        x if x >= 0.000_000_1 => (raw_value * 1_000_000.0, -2),
-        x if x >= 0.000_000_000_1 => (raw_value * 1_000_000_000.0, -3),
-        _ => (raw_value * 1_000_000_000_000.0, -4),
-    };
-
-    //println!("Magnitude: ({}, {})", magnitude_value, magnitude_level);
-
-    let (value, level) = if magnitude_level < min_unit_level {
-        (min_unit_value, min_unit_level)
-    } else {
-        (magnitude_value, magnitude_level)
-    };
+    let exp_level = (raw_value.log10().div_euclid(3.) as i32)
+        .max(min_exp_level)
+        .min(4);
```

Comment:
By the way, `clamp()` [is stabilized](https://github.com/rust-lang/rust/pull/77872) in [rust 1.50.0](https://github.com/rust-lang/rust/blob/master/RELEASES.md) which releases soon.